### ISSUE TITLE
Document that GET handler also handles HEAD requests

### ISSUE THREAD

**Issue by syrnick** [expressjs/expressjs.com#748]
The current docs don't mention that GET handlers must also accept HEAD requests:

https://github.com/expressjs/express/blob/master/lib/router/route.js#L65

This is quite unexpected and would cause bugs if something is done for GETs, but not for HEADs. E.g.
```
app.get('/', (req, res, next) => {
   assert(req.method === 'GET');
   ...
});
```
looks reasonable, but will throw for a HEAD request.

(Originally incorrectly posted to https://github.com/expressjs/express/issues/3149)

**Comment by crandmck** [comment]
OK, if I understand what you're saying, we should add a note saying that `app.get()` and `router.get()` will handle HTTP HEAD requests as well as GET requests.  Is that right?

[According to the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.4):
> ... the server MUST NOT return a message-body in the response.

So perhaps we should note that as well?

Ideally we would provide a short example of the right way to handle both GET and HEAD.  @dougwilson Is that just a matter of checking `req.method` and doing the appropriate thing for `GET` vs. `HEAD`?

**Comment by dougwilson** [comment]
Express handles the details for the users. Users are expected to just do the GET action in their GET route and not look at the req.method. This is because per spec a HEAD is expected to perform the same action as a GET for the same path, but simply not return a body. Express (and Node.js as well) ensure that the body is dropped automatically.

**Comment by syrnick** [comment]
It just needs a note like this: A GET handler MUST handle HEAD requests unless a HEAD handler was explicitly added. 

@dougwilson the spec says nothing about the "action" it just talks about the headers. E.g. if a GET request would do some work, warm up caches and load things into memory, HEAD request might do none of that. It could only check permissions and last modified time and be done.

I think it's a good idea to call GET handlers for HEAD requests. Developers just need a reminder that HEAD requests exist and they should expect them in GET handlers. If they call a function that is available only for the GET requests (e.g. provided by some middleware), they should check for request type first. 

The following will look crazy to many and they'd need to learn about https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.4 and  https://github.com/expressjs/express/blob/master/lib/router/route.js#L65 .

```
app.get('/', (req, res, next) => {
   if (req.method === 'GET') {
   } else {
   }
});
```

**Comment by dougwilson** [comment]
Yea, sorry, I just wrote that quickly on a phone :) So when I said spec, I really meant we are trying to adhere to as many "SHOULD" as possible out-of-the-box for users, especially since Express was originally developed as a general purpose webserver, vs a strict REST server. The doc you linked to said

> The metainformation contained in the HTTP headers in response to a HEAD request SHOULD be identical to the information sent in response to a GET request.

And so to achieve that, we are calling the defined get handler in Express and simply dropping the body. Many things in Express, like ETag, Content-Length, etc., need the body to calculate, so most people will need to retrieve the body / render the template no matter.

If people want to & can with their architecture do something to simplify HEAD metainformation lookup, they can always define a head handler for their route :)

**Comment by dougwilson** [comment]
@syrnick it may be even easier to get the docs updated by editing the site yourself, using the "Edit this page on GitHub" link that it on the bottom of every page. Then it can just be merged in vs someone needing to get around to making the change at some point :) It will also help us understand what wording you think is helpful or not instead of trying to figure it out here :)

**Comment by dougwilson** [comment]
I quickly put together a doc update proposal in #750

**Comment by crandmck** [comment]
Fixed by #750 ... thanks @dougwilson and @syrnick
